kernel/src/thread/task.rs

```diff
@@ -88,15 +90,71 @@ pub fn create_new_user_task(
             if current_thread.is_exited() {
                 break;
             }
-            handle_pending_signal(user_ctx, &ctx, syscall_number);
-            // If current is suspended, wait for a signal to wake up self
-            while current_thread.is_stopped() {
-                Thread::yield_now();
-                debug!("{} is suspended.", current_posix_thread.tid());
-                handle_pending_signal(user_ctx, &ctx, None);
+
+            enum Status {
+                /// The thread has exited.
+                Exit,
+                /// The thread is currently running.
+                Run,
+                /// The thread is stopped, but there are pending signals.
+                /// We need to handle the pending signals to determine if any can
+                /// wake up the thread.
+                RecheckSignal,
             }
-            if current_thread.is_exited() {
-                debug!("exit due to signal");
+
+            let mut status = Status::Run;
+
+            loop {
+                // Fast path: Check if the thread has exited or is not stopped
+                handle_pending_signal(user_ctx, &ctx, syscall_number);
+
+                if current_thread.is_exited() {
+                    status = Status::Exit;
+                    break;
+                }
+
+                if !current_posix_thread.status().lock().is_stopped() {
+                    break;
+                }
+
+                // Slow path: The thread is stopped.
+                let wait_status = stop_wait_queue.pause_until(|| {
+                    // FIXME: Should signals other than SIGCONT and SIGKILL be handled while the thread is stopped?
+                    handle_pending_signal(user_ctx, &ctx, syscall_number);
+
+                    if current_thread.is_exited() {
+                        return Some(Status::Exit);
+                    }
+
+                    // Check for pending signals before acquiring the thread status lock,
+                    // as `has_pending` may internally acquire a Mutex lock.
```

Comment:
`pause_until` already checks pending signals. I'm not sure why we need this check here?